### PR TITLE
Expose `arrowBoundaryOffset ` and internalize `arrowSize` calculation on RAC `<Popover />` and `<Tooltip />`

### LINKED ISSUE #5917: Tooltip: OverlayArrow placed too far on the edges
### Provide a general summary of the issue here

Screenshot 2024-02-21 at 14 21 17

### 🤔 Expected Behavior?

The OverlayArrow shouldn't go too far to the edges of the tooltip.

### 😯 Current Behavior

The OverlayArrow goes all the way to the edge when the trigger is wider than double the width of the tooltip. This is an issue when there's border radius applied to the tooltip. The positioning is set internally, not sure if it's from the TooltipTrigger or the Tooltip. 

### 💁 Possible Solution

Could you provide a prop to clamp the positioning of the OverlayArrow, so that it does not go further to the edge than allowed?

### 🔦 Context

_No response_

### 🖥️ Steps to Reproduce

1. Go to your [storybook example, with placement: top right](https://react-spectrum.adobe.com/react-aria-starter/index.html?path=/story/tooltip--example&args=placement:top+right)
2. Set the width of the trigger/button to a width wider than double the width of the tooltip (100px for example).

### Version

1.1.0

### What browsers are you seeing the problem on?

Chrome

### If other, please specify.

_No response_

### What operating system are you using?

Macos Sonoma

### 🧢 Your Company/Team

_No response_

### 🕷 Tracking Issue

_No response_

### PR DESCRIPTION
**Pull request by sookmax**
Closes #5917

Hello, I thought it would be nicer, in terms of DX, to add a prop with which users can adjust their arrow position, rather than exposing `OverlayArrowContext` as suggested in https://github.com/adobe/react-spectrum/issues/5917#issuecomment-1957976622.

The way this PR implements the feature relies on the fact that `calculatePosition` returns only one of either `arrowOffsetLeft` or `arrowOffsetTop` (not both) with a valid value. So I added a defensive test case in `calculatePosition.test.ts` as well.

Frankly, it's a little unintuitive which way is positive/negative when setting the `props.offset`, but I believe it's easy enough for the user to figure out the direction of the offset by changing it a few times.

Or maybe there's yet a better way to do this. I'm open to any suggestion from you guys!

Thanks.

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/WAI/ARIA/apg/)

### Stories
http://localhost:9003/?path=/story/react-aria-components--tooltip-arrow-offest-example&providerSwitcher-express=false&strict=true

### Docs
* http://localhost:1234/react-aria/Popover.html#overlayarrow
* http://localhost:1234/react-aria/Tooltip.html#overlayarrow

### CODE REVIEW

**Comment by snowystinger** [comment]
Hey! We had a chance to talk about this one today. Instead of adding a prop to the OverlayArrow, we'd like to expose arrowOffsetBoundary on Popover, which already exists as a prop on usePopover
(you can see how we use it in React Spectrum's implementation here https://github.com/adobe/react-spectrum/blob/a8ec9667a2b70b88b45e329bf44918f0f93d54f9/packages/%40react-spectrum/overlays/src/Popover.tsx#L116)
and can see the prop description here https://react-spectrum.adobe.com/react-aria/usePopover.html#usepopover

This will allow the value to get into calculatePosition reducing the work a user has to do.

In addition, we think that we could get a ref to the OverlayArrow through https://github.com/adobe/react-spectrum/blob/a8ec9667a2b70b88b45e329bf44918f0f93d54f9/packages/react-aria-components/src/Popover.tsx#L167 which would allow us to automatically measure the size of the arrow, negating the need to also expose that prop from Popover.

**Comment by thorbde** [comment]
@snowystinger You mentioned exposing the prop in Popover, would it also be exposed in Tooltip?

**Comment by sookmax** [comment]
@snowystinger Thanks for the feedback! I revised the fix according to your feedback to both `<Popover>` and `<Tooltip>`. `<Popover>` actually has already been exposing `arrowBoundaryOffset` so I just added a story showcasing the usage of that and also removed `props.arrowSize`.

[The usage of `Pick<>` here](https://github.com/adobe/react-spectrum/pull/5936/files#diff-292f27e58624fe5934f759580e686ca5b14da4dac87975fee81366fbeaabc6deR26) will properly be reflected in the docs once https://github.com/adobe/react-spectrum/pull/5986 gets merged.

Let me know if I missed anything!

**Comment by sookmax** [comment]
hm not sure why the test `packages/@react-aria/interactions/test/useFocusVisible.test.js` is failing on CI. It passes fine in my local env though 😭.

https://app.circleci.com/pipelines/github/adobe/react-spectrum/19178/workflows/a98494f0-8486-468b-b927-28aac3a1c025/jobs/278905

**Comment by snowystinger** [comment]
GET_BUILD

File: packages/react-aria-components/stories/Popover.stories.tsx [hunk 1]
```diff
@@ -48,3 +48,218 @@ export const PopoverExample = () => (
...
+export const PopoverArrowBoundaryOffsetExample = () => {
+  return (
+    <div style={{display: 'flex', flexDirection: 'column'}}>
+      <div style={{display: 'flex'}}>
+        <div style={{padding: 12}}>
+          <DialogTrigger>
+            <Button style={{width: 200, height: 100}}>Top left</Button>
+            <Popover
+              placement="top left"
+              arrowBoundaryOffset={25}
```
Comment (LFDanLu) [inline packages/react-aria-components/stories/Popover.stories.tsx:61]: Just one suggestion, mind adding a storybook control for this value so we can check how it behaves at extreme values?

Comment (sookmax) [reply packages/react-aria-components/stories/Popover.stories.tsx]: @LFDanLu thanks for the suggestion! I added what you asked for.


https://github.com/adobe/react-spectrum/assets/71210554/4a77434a-51df-413f-960c-14415ad6fb6a

**Comment by snowystinger** [comment]
GET_BUILD

**Review by snowystinger — APPROVED** [review]
It looks like we need to do a little more clamping work as I can get them to separate, and it also affects the overall placement when I set the offsetboundary to something extreme. I think we can handle it in a follow up
Screenshot 2024-03-11 at 8 08 20 AM
it's probably not a common case. most of the time this number should be the border radius of the popover, in which case the most it'll ever be is 50% of the height where we don't need to worry about clamping yet.

Thanks for the stories and PR!

**Pull request merged**